plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixPageSource.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.phoenix5;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RowBlock;
+import io.trino.spi.connector.ConnectorPageSource;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+public class PhoenixPageSource
+        implements ConnectorPageSource
+{
+    private final ConnectorPageSource delegate;
+    private final List<Integer> mergeRowIdSourceChannels;
+    private final int mergeRowIdChannel;
+    private final int columnCount;
+
+    public PhoenixPageSource(ConnectorPageSource delegate, List<Integer> mergeRowIdSourceChannels, int mergeRowIdChannel, int columnCount)
+    {
+        this.delegate = delegate;
+        this.mergeRowIdSourceChannels = mergeRowIdSourceChannels;
+        this.mergeRowIdChannel = mergeRowIdChannel;
+        this.columnCount = columnCount;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page = delegate.getNextPage();
+        if (page == null) {
+            return null;
+        }
+        int positionCount = page.getPositionCount();
+
+        Block[] mergeRowIdBlocks = new Block[mergeRowIdSourceChannels.size()];
+        for (int i = 0; i < mergeRowIdBlocks.length; i++) {
+            mergeRowIdBlocks[i] = page.getBlock(mergeRowIdSourceChannels.get(i));
+        }
+        Block mergeRowId = RowBlock.fromFieldBlocks(positionCount, Optional.empty(), mergeRowIdBlocks);
+
+        Page returnPage = new Page(positionCount);
+        for (int channel = 0; channel < page.getChannelCount(); channel++) {
+            if (channel == mergeRowIdChannel) {
+                returnPage = returnPage.appendColumn(mergeRowId);
+            }
+            returnPage = returnPage.appendColumn(page.getBlock(channel));
```

Comment:
We should try avoiding use of appendColumn. This incurs arrayCopy on block references array. It's fine to do it once per page, but in this case we will incur the copy for every appendColumn call per page. So it's as many array copies as number of columns per page.